Initialization from rotation vector

adbugger opened the pull request:
Initial implementation and tests for initialization from rotation vector

adbugger committed:
ENH: Implemented initialization from rotation vector

adbugger committed:
TST: Added test cases for `from_rotvec` function

adbugger committed:
MAINT: Fixed implementation of from_rotvec

adbugger committed:
TST: Added identity quaternion in tests for `from_rotvec`

adbugger committed:
DOC: Added `from_rotvec` to documentation

adbugger committed:
MAINT: Fixed typo in function

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -195,3 +196,58 @@ def from_dcm(cls, dcm):
+
+    @classmethod
+    def from_rotvec(cls, rot_vecs):
+        """Initialize class from rotation vector.
```
rot_vecs -> rotvec for uniformity?

nmayorov replied:
rot_vecs -> rotvec for uniformity?

Apply everywhere.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -195,3 +196,58 @@ def from_dcm(cls, dcm):
+
+        quat = np.empty((num_rotations, 4))
+
+        zero_norms = np.nonzero(norms == 0)[0]
```
No reason to consider this case separately from small_angle case.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -195,3 +196,58 @@ def from_dcm(cls, dcm):
+
+        small_angle = np.nonzero(np.logical_and(
+            norms <= 1e-3, norms != 0))[0]
+        quat[small_angle, :3] = (0.5 - np.power(norms[small_angle], 2) /
```
Don't know how much this is an improvement, but maybe introduce `scale = np.empty(n_rotations)`. Then `scale[small_angle] = ...` and `scale[large_angle] = ...` and finally `quat[:, :3] = scale[:, None] * rotvec`?

This way you also don't need to broadcast `norms` to 2d which is a slight improvement in clarity as well.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -195,3 +196,58 @@ def from_dcm(cls, dcm):
+
+        small_angle = np.nonzero(np.logical_and(
+            norms <= 1e-3, norms != 0))[0]
+        quat[small_angle, :3] = (0.5 - np.power(norms[small_angle], 2) /
```
Why do you prefer to use `power` instead of `**`?

adbugger replied:
In this particular case, I read that the results of `np.power` and `**` can sometimes differ, not enough to be seen by `np.allclose` where tolerance is `1e-8` by default. Since I'm not sure how much of a concern this is, I'll use `**`.

nmayorov commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -200,3 +200,35 @@ def test_from_dcm_ortho_output():
+    assert_array_almost_equal(result._quat, expected_quat)
+
+
+def test_from_generic_rotvec():
```
Add some tests with small (but non zero) angles as well. Think how to test it. 

My ideas is that `w` should be close to 1 (up to quadratic term) and `x, y, z` should be approximately `0.5 * rotvec`. Also check that the norm is very close to 1.

nmayorov reviewed (commented):
Decent implementation, I suggested some possible improvements.

adbugger committed:
MAINT: Renamed variable and used ** instead of np.power

larsoner commented:
@adbugger let me know when @nmayorov is happy and I'll take a look, too

adbugger committed:
MAINT: Changed implementation for clarity

Introduce `scale` variable in `from_rotvec` for clarity

adbugger committed:
TST: Added new tests and modified old ones for from_rotvec

Added a test case for small angles. Modified previous tests to use
as_quaternion function instead of accessing ._quat directly

adbugger committed:
STY: Removed redefinition of `small_norms` variable

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -195,3 +196,56 @@ def from_dcm(cls, dcm):
+        large_angle = np.nonzero(norms > 1e-3)[0]
+        scale[large_angle] = (np.sin(norms[large_angle] / 2) /
+                              norms[large_angle])
+
```
Let's create `quat` here.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -195,3 +196,56 @@ def from_dcm(cls, dcm):
+
+        num_rotations = rotvec.shape[0]
+
+        norms = np.linalg.norm(rotvec, axis=1)
```
I suggest to group `norms`, `small_angle` and `large_angle` in consecutive lines.

And then group `scale` and its element assignment.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -195,3 +196,56 @@ def from_dcm(cls, dcm):
+
+        scale = np.empty(num_rotations)
+
+        small_angle = np.nonzero(np.logical_and(
```
Don't need to add `norm != 0` here.

nmayorov commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -200,3 +201,57 @@ def test_from_dcm_ortho_output():
+
+def test_from_rotvec_small_angle():
+    # Small theta
+    rotvec = np.array([1, 0.00001, 0.00001]) * 5e-6
```
The axis is not generic enough (along `x` practically) and I don't understand why you defined it this way. Could you test with a more interesting axis and perhaps somewhat lager angle?

Also we probably want to test on a stack of vectors where large and small angles are present.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -195,3 +196,56 @@ def from_dcm(cls, dcm):
+                              norms[large_angle])
+
+        quat[:, :3] = scale[:, None] * rotvec
+        quat[:, 3] = np.cos(norms / 2).flatten()
```
Don't need to to flat it here.

nmayorov reviewed (commented):
Some suggestions on how to reformat code to make it easier to follow.

Also try to improve tests for small angles.

adbugger committed:
STY: Style changes to code

adbugger committed:
TST: Better testing for small angles

adbugger committed:
STY: Removed extra blank line

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -195,3 +196,54 @@ def from_dcm(cls, dcm):
+        norms = np.linalg.norm(rotvec, axis=1)
+        small_angle = np.nonzero(norms <= 1e-3)[0]
+        large_angle = np.nonzero(norms > 1e-3)[0]
+
```
Move `scale` here and I will be happy.

adbugger replied:
I did exactly that in the latest commit :P

nmayorov commented:
It looks good.

I have one idea about testing. My concern is that your tests contain some hard coded specific numbers, which is not necessary bad, but not ideal I guess.

We can test that the conversion is correct, if --- `w = cos(r / 2)` and `(x, y, z) is along r_` (can check with cross product) and `x**2 + y**2 + z**2 + w**2 = 1`. I don't suggest to rewrite your tests like this, unless you really like this idea. Maybe postpone if for later, when we have `to_rotvec`.

@larsoner also take a look.

@adbugger start working on `to_rotvec`.

nmayorov commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -200,3 +201,71 @@ def test_from_dcm_ortho_output():
+    assert_allclose(
+            quat[1, :3],
+            np.array([
+                0.098796039321534656925304764958285802128222305393864012037,
```
Shorten these numbers somehow, 17 digits should be well enough.

adbugger committed:
TST: Shortened numbers in tests

larsoner commented on scipy/spatial/transform/rotation.py:
```diff
@@ -195,3 +196,52 @@ def from_dcm(cls, dcm):
+        num_rotations = rotvec.shape[0]
+
+        norms = np.linalg.norm(rotvec, axis=1)
+        small_angle = np.nonzero(norms <= 1e-3)[0]
```
no need for `np.nonzero` just use boolean masks directly. Then you can do:
```
small_angle = (norms <= 1e-3)
large_angle = ~small_angle
```
And the rest of the code should work

larsoner commented on scipy/spatial/transform/rotation.py:
```diff
@@ -195,3 +196,52 @@ def from_dcm(cls, dcm):
+        large_angle = np.nonzero(norms > 1e-3)[0]
+
+        scale = np.empty(num_rotations)
+        scale[small_angle] = (0.5 - norms[small_angle] ** 2 / 48 +
```
A comment with a name or ref for the technique used here (taylor expansion of sin?) would be good. Think about the next dev who has to work with this code, they will probably not immediately know what this is doing

larsoner commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -200,3 +201,71 @@ def test_from_dcm_ortho_output():
+
+
+def test_malformed_1d_from_rotvec():
+    with pytest.raises(ValueError):
```
In general it's good practice to use `match='at least a few words from the error message'` here to ensure you catch the right error

larsoner reviewed (commented):
Other than a few minor gripes, LGTM.

adbugger committed:
MAINT: Used boolean masks instead of np.nonzero

adbugger committed:
TST: Added match parameter to pytest.raises(ValueError)

adbugger commented:
Should I merge this? The failures say internal error, I'm not sure what that means.

adbugger commented:
Merging this due to review approval. I'm going to rebase `rotation_as_rotvec` off this, and open another pull request.
If required, I'll open another PR later to make fixes to this.
